### PR TITLE
Add more options other than "spectators can attend for free"

### PR DESCRIPTION
**Pull request by notrichardpeng**
Addresses #6321 

Right now the idea is to have a selection menu pop up in the competition form when the entry fee for guests is set to 0.

But in Javascript I'm breaking the design pattern currently, because `.on('change')` to the element `name=competition[guests_entry_fee_lowest_denomination]` can only be detected when the page is loaded up initially. So, I had to use the other input field of the whole currency input div, that didn't have a `name` attribute before I added it myself. However I couldn't figure out how to add the `competition[]` bracket around `guests_entry_fee_lowest_denomination` around its name to keep things consistent.

### CODE REVIEW

File: WcaOnRails/config/locales/en.yml
Comment (gregorbg) [inline WcaOnRails/config/locales/en.yml:635]: Quotes are not necessary here, the above `"true"` and `"false"` are quoted because normally they would be boolean keywords.

File: WcaOnRails/app/inputs/money_amount_input.rb
Comment (gregorbg) [inline WcaOnRails/app/inputs/money_amount_input.rb:33]: What is this change about?

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:912]: Try to define these as enum. You can take `stripe_charge.rb` as an example.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:916]: enum, see my previous comment.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [reply WcaOnRails/app/models/competition.rb:912]: This is an article I found helpful when researching about Rails enums: https://betterprogramming.pub/how-to-use-enums-in-rails-6-87600e292476

File: WcaOnRails/app/views/competitions/_competition_form.html.erb
Comment (gregorbg) [inline WcaOnRails/app/views/competitions/_competition_form.html.erb:188]: Maybe (I'm not sure myself, you will have to try) you can re-use the enum that I suggested above and pass it to the `collection` parameter here. (Or perhaps a `collection` parameter isn't even necessary anymore when Ruby detects that the `:free_entry_text` field is backed by an enum)

File: WcaOnRails/app/views/competitions/_registration_requirements.html.erb
Comment (gregorbg) [inline WcaOnRails/app/views/competitions/_registration_requirements.html.erb:107]: You can move the `<br>` out of the `if` query:

```
<% if foo %>
  <%= t() %>
<% elsif bar %>
  <%= t() %>
<% end %>
<br>
```

File: WcaOnRails/db/migrate/20210721135845_add_free_entry_text_to_competitions.rb
Comment (gregorbg) [inline WcaOnRails/db/migrate/20210721135845_add_free_entry_text_to_competitions.rb:5]: Why allow `NULL` values if you provide a default anyways?

**Review by gregorbg — COMMENTED** [review]
I'm not sure about calling the column `free_entry_text`. Effectively it seems by looking at the code that the organiser is choosing between three different "spectator admission" modes, but they cannot enter any free form text by themselves. You are deriving a pre-defined text based on which "mode" the organiser chose.

To be clear, I like that approach (not storing free text sentences in our DB is good!) but you should think of a name for the column that reflects the circumstances that I described above.

File: WcaOnRails/app/inputs/money_amount_input.rb
Comment (notrichardpeng) [reply WcaOnRails/app/inputs/money_amount_input.rb:33]: So I was looking for a way to make the selection menu (the radio buttons) for free entry display appear immediately when `Entry fee for guests` on the competition creation page is set to zero. 

The `Entry fee for guests` text field has two `input` fields in HTML, but only one of them has a `name` attribute for `$('input[name=""]').on('change')` in javascript.  But the one that has the `name` attribute does not update its `value` attribute in real-time (I think it does update in HTML, but it couldn't be detected by javascript immediately). 

However, the second `input` element of the text field _does_ update its `value` in real-time (can be detected by javascript), but it doesn't have a `name` attribute so I found this way to give it a `name`. This is also why the javascript I added is a little inconsistent with the other line of code, because the `name` of this element does not have `competitions[]` in front of it. 

Not sure if this was the best solution though, since this change affects other money_amount_input elements too. What are your thoughts?

File: WcaOnRails/app/models/competition.rb
Comment (notrichardpeng) [reply WcaOnRails/app/models/competition.rb:912]: Enums! Haven't used them in so long that I forgot they exist.

File: WcaOnRails/db/migrate/20210721135845_add_free_entry_text_to_competitions.rb
Comment (notrichardpeng) [reply WcaOnRails/db/migrate/20210721135845_add_free_entry_text_to_competitions.rb:5]: oops

File: WcaOnRails/app/views/competitions/_competition_form.html.erb
Comment (notrichardpeng) [reply WcaOnRails/app/views/competitions/_competition_form.html.erb]: Yep it worked, luckily sombody installed `enum_helper` before me

File: WcaOnRails/db/migrate/20210801044007_add_free_entry_status_to_competitions.rb
Comment (gregorbg) [inline WcaOnRails/db/migrate/20210801044007_add_free_entry_status_to_competitions.rb:5]: Please be explicit about not allowing `NULL`.

File: WcaOnRails/config/locales/en.yml
Comment (gregorbg) [inline WcaOnRails/config/locales/en.yml:434]: Nit about the name of this enum, see above.

File: WcaOnRails/config/locales/en.yml
Comment (gregorbg) [inline WcaOnRails/config/locales/en.yml:312]: Nit about the name of this enum, see above.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:79]: Great! Also +1 for using an explicit hash and not relying on implicit array ordering.
I'd argue that this field should be called something like `free_guest_entry_status` because it does not affect the competitor entry fee.

File: WcaOnRails/app/inputs/money_amount_input.rb
Comment (gregorbg) [reply WcaOnRails/app/inputs/money_amount_input.rb:33]: Yo no hablo jQuery. Will need feedback by other WST members on this.

**Review by gregorbg — CHANGES_REQUESTED** [review]
Thanks for the prompt fixes! Some remaining nitpicks but generally looks much cleaner now :D

Also, no clue about what the "best" approach for updating the `Entry fee for guests` field would be. I happily defer to other WST members' experience for this.

File: WcaOnRails/app/models/competition.rb
Comment (notrichardpeng) [reply WcaOnRails/app/models/competition.rb:79]: :o indeed

File: WcaOnRails/app/inputs/money_amount_input.rb
Comment (notrichardpeng) [reply WcaOnRails/app/inputs/money_amount_input.rb:33]: No hablo jQuery tambien!

**Comment by notrichardpeng** [comment]
Kind of new to migrating the database, every time I do `bin/rails db:migrate`  it changes a bunch of other lines. For example it changes comments like `"!40101 SET character_set_clien = xxx"` to `"!5050s SET character_set_client = utf8mb4"`, and it also takes away brackets from type definitions such as from `int(11) DEFAULT NULL` to `int DEFAULT NULL`. Should I revert these changes every time I do db:migrate?

File: WcaOnRails/app/inputs/money_amount_input.rb
Comment (jonatanklosko) [reply WcaOnRails/app/inputs/money_amount_input.rb:33]: The money input consists of a visible nameless input and a hidden one with `name=competition[guests_entry_fee_lowest_denomination]`. Whenever the visible one changes, the former value is synchronized, and that's the one that gets submitted with the form. Subscribing to the hidden input with `.on('change', ...)` doesn't work because it's value is set programatically and there is no event. What I'd try is modifying this:

https://github.com/thewca/worldcubeassociation.org/blob/fec596a07772f315a85c4dd9f7c688d9daa4709d/WcaOnRails/app/webpacker/javascript/auto-numeric/index.js#L74

To `.val(...).trigger('change');`. With that in place, you should be able to subscribe to the hidden input changes and keep the visible one nameless.

**Review by gregorbg — COMMENTED** [review]
LGTM pending the one change that Jonatan suggested via comment.

After some googling, it seems like the `structure.sql` changes heavily depend on which exact MySQL version you're dumping the schema with. (see also https://stackoverflow.com/questions/52989311/search-path-in-structure-sql-file-gets-modified-by-rails-dbmigrate)
As we are using wildly different versions please only commit the lines that are actually relevant to your change (in this case should just be one, so it's fine as-is).
I will think of a more permanent solution to this problem later.

File: WcaOnRails/app/inputs/money_amount_input.rb
Comment (notrichardpeng) [reply WcaOnRails/app/inputs/money_amount_input.rb:33]: Gotya, thank you Jonatan!

**Comment by notrichardpeng** [comment]
Can I merge this?

**Comment by gregorbg** [comment]
I would like to use this opportunity for you to practice writing RSpec tests.

In particular, this new feature adds a new UI element which I believe warrants additional tests. At least two tests in `features/competition_management_spec`, one that sets the guest entry fee to 0 and expects the new form input to appear, another that sets it to anything != 0 and expects the new form input to be absent.
Perhaps you could also add a test to `models/competition_spec` that creates a mock competition for each new guest entry status and then checks the `all_guests_allowed?` and `some_guests_allowed?` properties.
And of course whatever you can come up with yourself!

**Comment by notrichardpeng** [comment]
I'll be looking at other tests tomorrow

**Comment by notrichardpeng** [comment]
Something interesting I noticed locally... whenever there are trailing whitespace in `competition_management_spec.rb`, javascript becomes not responsive and causes `change guest entry fee to non-zero` scenario to fail, where the selection menu does not disappear even though the entry fee for guests is set to non-zero.

**Review by gregorbg — APPROVED** [review]
LGTM, thank you for adding the tests :+1:

**Pull request merged**
